cli/src/cli_util.rs

```diff
@@ -2732,28 +2742,47 @@ fn resolve_aliases(
                         r#"Recursive alias definition involving "{alias_name}""#
                     )));
                 }
-                if let Some(value) = aliases_map.remove(&alias_name) {
-                    if let Ok(alias_definition) = value.try_deserialize::<Vec<String>>() {
-                        assert!(string_args.ends_with(&alias_args));
-                        string_args.truncate(string_args.len() - 1 - alias_args.len());
-                        string_args.extend(alias_definition);
-                        string_args.extend_from_slice(&alias_args);
-                        resolved_aliases.insert(alias_name.clone());
-                        continue;
-                    } else {
-                        return Err(user_error(format!(
-                            r#"Alias definition for "{alias_name}" must be a string list"#
-                        )));
-                    }
+                if let Some(alias_definition) = resolved_alias_map.get(&alias_name) {
+                    assert!(string_args.ends_with(&alias_args));
+                    string_args.truncate(string_args.len() - 1 - alias_args.len());
+                    string_args.extend(alias_definition.clone());
+                    string_args.extend_from_slice(&alias_args);
+                    resolved_aliases.insert(alias_name.clone());
+                    continue;
                 } else {
                     // Not a real command and not an alias, so return what we've resolved so far
                     return Ok(string_args);
                 }
             }
         }
         // No more alias commands, or hit unknown option
-        return Ok(string_args);
+        break;
     }
+
+    // Add each alias to `Command` so it shows up in help
+    for (alias_name, alias_definition) in resolved_alias_map {
+        // Find the innermost subcommand so we can use its help.
+        let mut subcmd = &*app;
+        for arg in &alias_definition {
+            if let Some(cmd) = subcmd.find_subcommand(arg) {
+                subcmd = cmd;
+            } else {
+                break;
+            }
+        }
+        let help = format!("Alias for {alias_definition:?}");
+        let subcmd_alias = subcmd
+            .clone()
+            .name(&alias_name)
+            // if we leave the built-in aliases, clap doesn't know whether to call the user alias or
+            // the built-in command
+            .alias(None)
+            .before_help(&help)
+            .before_long_help(help);
+        *app = std::mem::take(app).subcommand(subcmd_alias);
```

Comment:
Aside: I'm at least going to experiment with a `&mut` builder API in clap v5.  I think it could simplify some code and I have a unsubstantiated hope it can help with binary sizes.